query-engine/core/src/query_graph_builder/write/upsert.rs

```diff
@@ -58,35 +57,50 @@ pub fn upsert_record(
     model: ModelRef,
     mut field: ParsedField,
 ) -> QueryGraphBuilderResult<()> {
-    graph.flag_transactional();
+    let where_argument = field.where_args().unwrap()?;
+    let create_argument = field.create_args().unwrap()?;
+    let update_argument = field.update_args().unwrap()?;
 
-    let where_arg: ParsedInputMap = field.arguments.lookup(args::WHERE).unwrap().value.try_into()?;
+    let filter = extract_unique_filter(where_argument.clone(), &model)?;
+    let read_query = read::find_unique(field.clone(), model.clone())?;
 
-    let filter = extract_unique_filter(where_arg, &model)?;
-    let model_id = model.primary_identifier();
+    if can_use_connector_native_upsert(
+        &model,
+        &where_argument,
+        &create_argument,
+        &update_argument,
+        connector_ctx,
+    ) {
```

Comment:
There is still a clone that can be avoided here. The reason I suggested to compute `can_use_connector` before-hand is to prevent the compiler to cry because we're borrowing a moved value.

```rust
let can_use_native_upsert = can_use_connector_native_upsert(
  &model,
  &where_argument,
  &create_argument,
  &update_argument,
  connector_ctx,
);

let filter = extract_unique_filter(where_argument, &model)?;

if can_use_native_upsert {} else {}
```